File: simulation.m
function simulation
    
    function [dx, dy] = computeForce(x, y)
        %   dx_i = sum_{j != i} f(r_ij)*(x_i - x_j)
        %   dy_i = sum_{j != i} f(r_ij)*(y_i - y_j)
        %   where r_ij = distance between cell i and cell j in 2D,
        %   and x,y are each N x 1 with N cells in 2D.
        
        % Define Morse parameters
        A = 0.5;       % amplitude of repulsion
        alpha = 0.5; % decay rate of repulsion
        B = 1;       % amplitude of attraction
        beta = 2;  % decay rate of attraction
        
        morse = @(r) A*exp(-alpha*r) - B*exp(-beta*r);

        c = 1;
        d = 1;
        h = 10;
        f = @(r) c(1-r/h);
        g = @(r) d(1-r/h);

    
        N = length(x);
        dx = zeros(N,1);
        dy = zeros(N,1);
    
        for i = 1:N
            for j = 1:N
                if j ~= i
                    % Differences in x,y
                    diffx = x(i) - x(j);
                    diffy = y(i) - y(j);
    
                    % Euclidean distance between (x_i,y_i) and (x_j,y_j)
                    r = sqrt(diffx^2 + diffy^2);
    
                    % Optionally check r>0 to avoid dividing by zero
                    % if your f(r) needs that. Here it's not shown.
                    val = morse(r);
                    % val = f(r);
    
                    % Sum up contributions
                    dx(i) = dx(i) + val * diffx;
                    dy(i) = dy(i) + val * diffy;
                end
            end
        end
    end


    function locDensity = computeDensity(x, y)
        
        coords = [x, y];
        distMat = squareform(pdist(coords)); 

        % % WEIGHTED MEAN DIST
        % % Choose a bandwidth parameter h that defines the neighborhood size
        % h = 20; 
        % 
        % % Compute weights that decay with distance (Gaussian kernel)
        % w = exp(-((distMat/h).^2));
        % % w = exp(-distMat);
        % 
        % % Exclude self-contributions by zeroing the diagonal
        % w(eye(size(w))==1) = 0;
        % 
        % % Compute a weighted mean distance for each cell
        % weightedMeanDist = sum(distMat .* w, 2) ./ sum(w, 2);
        % % disp(weightedMeanDist)
        % 
        % % Define local density as the inverse of the weighted mean distance
        % % locDensity = 1 ./ weightedMeanDist;
        % locDensity = exp(-(weightedMeanDist/h).^2);
        % disp(locDensity)

        % % UNWEIGHTED MEAN DIST
        % N = size(coords,1);
        % % sum distance from each row i to all columns j
        % sumDist = sum(distMat, 2);
        % % average distance to the other cells
        % meanDist = sumDist / (N-1);
        % % Define local density as inverse of mean distance
        % locDensity = 1 ./ meanDist;

        % locDensity = (locDensity - min(locDensity)) / (max(locDensity) - min(locDensity));
        % disp(locDensity)

        % FIXED RADIUS NEIGHBORS
        r0 = 15;

        % Create a logical matrix indicating which distances are <= r0
        neighborMask = distMat <= r0;

        % Exclude self by setting the diagonal to false
        neighborMask(1:size(neighborMask,1)+1:end) = false;

        % Count the number of neighbors for each coordinate
        neighborCounts = sum(neighborMask, 2);

        h = 5; 
        locDensity = 1-exp(-neighborCounts./h);


        % disp(locDensity)
        % disp(neighborCounts)
        for i = 1:length(locDensity)
            fprintf('(%f, %d)\n', locDensity(i), neighborCounts(i));
        end
    end

    % Parameters
    numCells = 8;
    radius   = 100;
    numSteps = 250;
    
    % Allocate array to store positions (numCells x 2 x numSteps)
    % posHistory = zeros(numCells, 2, numSteps);
    history = cell(1, numSteps);
    
    % Initialize cell positions near center
    spawnRadius = 1;
    angles = 2*pi*rand(numCells,1);
    rDist  = spawnRadius * sqrt(rand(numCells,1));
    x = rDist .* cos(angles);
    y = rDist .* sin(angles);
    p = rand(numCells,1);
    d = computeDensity(x, y);
 
    
    % Store initial positions
    % posHistory(:,1,1) = x;
    % posHistory(:,2,1) = y;
    history{1} = [x, y, p, d];
    
    % Run the simulation in a loop, but NOT animating yet
    for t = 2:numSteps
        
        
        locDensity = computeDensity(x, y);
        
        lambda = 0.05;

        [dx, dy] = computeForce(x, y);
        dp = lambda*(1-locDensity);
        % disp(p)
        % disp(dp)
        x = x + dx;
        y = y + dy;
        p = p +dp;
        disp(t/numSteps)
        % disp(length(x))

        dividing = find(p >= 1);  % Indices of cells that reached p=1

        for idx = dividing'
            p(idx) = 0;           % Reset parent's p to 0
            epsDist = 0.75;        % small distance for new cell
            randAngle = 2*pi*rand;
    
            % compute new cell's position
            newX = x(idx) + epsDist*cos(randAngle);
            newY = y(idx) + epsDist*sin(randAngle);
    
            % Append to x, y, p
            x = [x;  newX];
            y = [y;  newY];
            p = [p;  0];         % new cell starts with p=0
        end
        
        % Enforce circular boundary
        distFromCenter = sqrt(x.^2 + y.^2);
        outside = distFromCenter > radius;
        if any(outside)
            scaleFactor   = radius ./ distFromCenter(outside);
            x(outside)    = x(outside) .* scaleFactor;
            y(outside)    = y(outside) .* scaleFactor;
        end

        d = computeDensity(x,y);
        % numCells = length(x);  % Update to include newly divided cells
        % posHistory(1:numCells,1,t) = x;
        % posHistory(1:numCells,2,t) = y;
        history{t} = [x, y, p, d];
    end
    
    % Figure
    hFig = figure('Name','Cell Simulation','NumberTitle','off');
    hAx  = axes('Parent', hFig);
    hold(hAx, 'on');
    axis(hAx, 'equal');
    caxis(hAx, [0, 1]);
    xlim(hAx, [-radius, radius]);
    ylim(hAx, [-radius, radius]);
    title(hAx, '2D Circle + Cells');
    
    % Draw boundary circle
    theta = linspace(0, 2*pi, 100);
    plot(hAx, radius*cos(theta), radius*sin(theta), 'k'); 
    
    % Plot initial positions
    scat = scatter(hAx, history{1}(:,1), history{1}(:,2), [], history{1}(:,3), 'filled');
    
    % Slider
    % Places a slider at the bottom of the figure
    % 'Min'=1, 'Max'=numSteps, start at 1
    hSlider = uicontrol('Style','slider', ...
        'Min',1, 'Max',numSteps, 'Value',1, ...
        'SliderStep',[1/(numSteps-1), 10/(numSteps-1)], ...
        'Units','normalized', 'Position',[0.1 0.02 0.6 0.05], ...
        'Callback', @sliderCallback);
    
    % Play/Pause
    % This button toggles a timer that automatically increments the slider
    hPlayPause = uicontrol('Style','pushbutton', 'String','Play', ...
        'Units','normalized', 'Position',[0.72 0.02 0.15 0.05], ...
        'Callback', @playPauseCallback);
    
    % Timer
    % We'll use a MATLAB timer object to periodically increment the slider
    playTimer = timer('ExecutionMode','fixedRate', ...
        'Period', 0.005, ...         % seconds between timer callbacks
        'TimerFcn', @autoAdvance); % function to call on each tick
   
    % Callback
    function sliderCallback(src, ~)
        % Called when the slider is moved
        t = round(get(src, 'Value'));  % current timestep
        updateScatter(t);
    end

    function playPauseCallback(src, ~)
        % Called when the "Play"/"Pause" button is clicked
        if strcmp(get(src, 'String'), 'Play')
            % Switch to "Pause" mode & start the timer
            set(src, 'String', 'Pause');
            start(playTimer);
        else
            % Switch to "Play" mode & stop the timer
            set(src, 'String', 'Play');
            stop(playTimer);
        end
    end

    function autoAdvance(~, ~)
        currentVal = round(get(hSlider, 'Value'));  % <-- round here
        if currentVal < numSteps
            newVal = currentVal + 1;
        else
            newVal = 1;  % loop back to start, or you could stop the timer
        end
        set(hSlider, 'Value', newVal);
        updateScatter(newVal);
    end

    function updateScatter(t)
        % Updates the scatter plot to show positions at time t
        data = history{t};
        xData = data(:,1);
        yData = data(:,2);
        cData = data(:,4);
        set(scat, 'XData', xData, 'YData', yData, 'CData', cData);
        drawnow;
    end

    % Cleanup
    % If the figure is closed while playing, stop the timer
    addlistener(hFig, 'ObjectBeingDestroyed', @(~,~) stopIfRunning());

    function stopIfRunning()
        if strcmp(playTimer.Running, 'on')
            stop(playTimer);
        end
        delete(playTimer);
    end

end

File: simulation.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.spatial.distance import pdist, squareform
from matplotlib.widgets import Slider, Button
import matplotlib.animation as animation
from IPython.display import HTML
import matplotlib.cm as cm
import matplotlib.colors as mcolors

# MARK: Functions

def morse(distances, A = 0.5, alpha = 0.5, B = 1, beta = 2):
    # A = 0.5, alpha = 0.5, B = 1, beta = 2
    # A = 2.5, alpha = 0.25, B = 5, beta = 1
    distances = np.asarray(distances)
    return A*np.exp(-alpha*distances)-B*np.exp(-beta*distances)

def exp_decay(distances, radius = 2):
    return np.exp(-distances/radius)

def normal(distances, magnitude = 0.5, radius = 0.2):
    return magnitude * np.exp(-((distances/radius)**2))

def linear(distances, radius = 20):
    values = distances*(-1/radius)+1
    return np.maximum(values, 0)

def density(x, y, radius = 15, threshold = 6):
    x = np.asarray(x)
    y = np.asarray(y)
    positions = np.stack((x, y), axis=1)
    dist_matrix = squareform(pdist(positions))
    neighbor_mask = (dist_matrix <= radius)
    np.fill_diagonal(neighbor_mask, False)
    neighbor_counts = np.sum(neighbor_mask, axis=1)

    # densities = 1 - np.exp(-neighbor_counts / threshold)
    densities = 1-linear(neighbor_counts, threshold)
    return densities


# MARK: Forces

def repulsion(x, y, lam=0.2, func=linear, radius=10):
    positions = np.column_stack((x, y))
    diffs = positions[:, np.newaxis, :] - positions[np.newaxis, :, :]
    dists = np.linalg.norm(diffs, axis=2)
    mask = ~np.eye(len(x), dtype=bool)

    weights = func(dists, radius)
    weights[~mask] = 0
    force_matrix = (weights[..., np.newaxis] * diffs).sum(axis=1)
    
    return lam * force_matrix[:, 0], lam * force_matrix[:, 1]

def vel_adhesion(x, y, dx, dy, lam=0.5, dist_func=linear, d_r=5):
    N = len(x)
    pos = np.column_stack((x, y))
    vel = np.column_stack((dx, dy))  # shape (N, 2)
    # Pairwise differences and distances
    diffs = pos[:, np.newaxis, :] - pos[np.newaxis, :, :]  # shape (N, N, 2)
    dists = np.linalg.norm(diffs, axis=2)                  # shape (N, N)
    # Get distance-based weights
    weights = dist_func(dists, d_r)                        # shape (N, N)
    # Zero out self-contributions
    np.fill_diagonal(weights, 0)
    # Multiply each neighbor's velocity by its weight
    weighted_velocities = weights @ vel  # matrix multiplication (N, N) x (N, 2) → (N, 2)
    # Scale by lambda
    return lam * weighted_velocities[:, 0], lam * weighted_velocities[:, 1]

def dist_adhesion(x, y, lam=0.5, dist_func=linear, d_r=5):
    N = len(x)
    pos = np.column_stack((x, y))

    # Compute pairwise position differences and distances
    diffs = pos[:, np.newaxis, :] - pos[np.newaxis, :, :]  # shape (N, N, 2)
    dists = np.linalg.norm(diffs, axis=2)                  # shape (N, N)

    # Get distance-based weights
    weights = dist_func(dists, d_r)                        # shape (N, N)
    np.fill_diagonal(weights, 0)  # no self-influence

    # Compute unit vectors toward each neighbor (avoid div-by-zero)
    with np.errstate(divide='ignore', invalid='ignore'):
        unit_diffs = np.divide(diffs, dists[..., np.newaxis], where=dists[..., np.newaxis] != 0)

    # Multiply weights with unit vectors and sum over neighbors
    forces = (weights[..., np.newaxis] * -unit_diffs).sum(axis=1)  # sum over neighbors

    return lam * forces[:, 0], lam * forces[:, 1]

def active(dx, dy, alpha):
    dx = np.asarray(dx)
    dy = np.asarray(dy)
    return alpha* dx, alpha*dy

def friction(dx, dy, gamma):
    dx = np.asarray(dx)
    dy = np.asarray(dy)
    v_mag_sq = dx**2 + dy**2
    return -gamma * v_mag_sq * dx, -gamma * v_mag_sq * dy
    

def cell_cycle(x, y, h=0.1):
    return h * (1 - density(x, y))


# MARK: Simulation
radius = 100
steps = 200

def init(num_cells = 8, spawn_radius = 1):
    angles = 2 * np.pi * np.random.rand(num_cells)
    r_dist = spawn_radius * np.sqrt(np.random.rand(num_cells))
    x = r_dist * np.cos(angles)
    y = r_dist * np.sin(angles)
    p = np.random.rand(num_cells)
    d = density(x, y)
    dx = np.zeros_like(x)
    dy = np.zeros_like(y)
    ddx = np.zeros_like(dx)
    ddy = np.zeros_like(dy)
    dp = np.zeros_like(p)
    return x, y, dx, dy, ddx, ddy, p, dp, d

def simulate(x, y, dx, dy, p, dp):
    # Compute forces
    # ad_ddx, ad_ddy = dist_adhesion(x, y, 0.2, linear, 10)
    ad_ddx, ad_ddy = vel_adhesion(x, y, dx, dy, 0.3, linear, 10)
    r_ddx, r_ddy = repulsion(x, y, 0.3, linear, 10)
    ac_ddx, ac_ddy = active(dx, dy, 0.1)
    f_ddx, f_ddy = friction(dx, dy, 0.1)

    ddx = ad_ddx + r_ddx + ac_ddx + f_ddx
    ddy = ad_ddy + r_ddy + ac_ddy + f_ddy

    dx += ddx
    dy += ddy

    # Cap velocity magnitude
    max_speed = 2.0  # adjust as needed
    speed = np.sqrt(dx**2 + dy**2)
    too_fast = speed > max_speed

    # Scale down dx and dy for fast cells
    dx[too_fast] *= max_speed / speed[too_fast]
    dy[too_fast] *= max_speed / speed[too_fast]

    x += dx
    y += dy

    dp = cell_cycle(x, y)
    p += dp

    dividing = np.where(p >= 1)[0]
    new_x = []
    new_y = []
    new_dx = []
    new_dy = []
    new_ddx = []
    new_ddy = []
    new_p = []
    new_dp = []

    for idx in dividing:
        p[idx] = 0
        dp[idx] = 0

        eps_dist = 1
        rand_angle = 2 * np.pi * np.random.rand()
        offset = eps_dist * np.array([np.cos(rand_angle), np.sin(rand_angle)])

        new_x.append(x[idx] + offset[0])
        new_y.append(y[idx] + offset[1])
        new_dx.append(dx[idx])
        new_dy.append(dy[idx])
        new_ddx.append(ddx[idx])
        new_ddy.append(ddy[idx])
        new_p.append(0.0)
        new_dp.append(0.0)

    if new_x:
        x = np.concatenate([x, new_x])
        y = np.concatenate([y, new_y])
        dx = np.concatenate([dx, new_dx])
        dy = np.concatenate([dy, new_dy])
        ddx = np.concatenate([ddx, new_ddx])
        ddy = np.concatenate([ddy, new_ddy])
        p = np.concatenate([p, new_p])
        dp = np.concatenate([dp, new_dp])

    # Enforce boundary via elastic collisions
    dist_from_center = np.sqrt(x**2 + y**2)
    outside = dist_from_center > radius

    if np.any(outside):
        # Normalize radial direction vectors
        nx = x[outside] / dist_from_center[outside]
        ny = y[outside] / dist_from_center[outside]

        # Radial component of velocity (dot product with normal)
        v_radial = dx[outside] * nx + dy[outside] * ny

        # Reflect: subtract 2 * (v · n) * n from velocity
        dx[outside] -= 2 * v_radial * nx
        dy[outside] -= 2 * v_radial * ny

        # Move particles just inside the boundary to prevent sticking
        x[outside] = nx * (radius - 1e-3)
        y[outside] = ny * (radius - 1e-3)


    d = density(x, y)
    return x, y, dx, dy, p, dp, d


x, y, dx, dy, ddx, ddy, p, dp, d = init()
history = []
history.append(np.column_stack((x, y, dx, dy, p, dp, d)))
for i in range(steps):
    print(i/steps)
    print(len(x))
    x, y, dx, dy, p, dp, d = simulate(x, y, dx, dy, p, dp)
    history.append(np.column_stack((x, y, dx, dy, p, dp, d)))

# MARK: GUI

def radial_colors(dx, dy):
    angles = np.arctan2(dy, dx)  # range [-π, π]
    hues = (angles % (2 * np.pi)) / (2 * np.pi)  # normalize to [0,1)
    # Create HSV color array and convert to RGB
    hsv_colors = np.zeros((len(hues), 3))
    hsv_colors[:, 0] = hues        # hue
    hsv_colors[:, 1] = 1.0         # full saturation
    hsv_colors[:, 2] = 1.0         # full brightness
    return mcolors.hsv_to_rgb(hsv_colors)

from matplotlib.colors import LinearSegmentedColormap, TwoSlopeNorm

# Custom diverging colormap: blue → black → red
black_center_cmap = LinearSegmentedColormap.from_list(
    'blue_black_red',
    [(0, 'blue'), (0.5, 'black'), (1, 'red')]
)

def skew_colors(x, y, dx, dy, cmap=black_center_cmap):
    r = np.column_stack((x, y))
    v = np.column_stack((dx, dy))

    r_norm = np.linalg.norm(r, axis=1, keepdims=True)
    v_norm = np.linalg.norm(v, axis=1, keepdims=True)
    r_unit = np.divide(r, r_norm, out=np.zeros_like(r), where=r_norm != 0)
    v_unit = np.divide(v, v_norm, out=np.zeros_like(v), where=v_norm != 0)

    inward = -r_unit

    dot = np.sum(inward * v_unit, axis=1)
    cross = inward[:, 0] * v_unit[:, 1] - inward[:, 1] * v_unit[:, 0]
    angles = np.arctan2(cross, dot)

    norm_angles = angles / np.pi  # range [-1, 1]

    norm = TwoSlopeNorm(vmin=-1, vcenter=0, vmax=1)
    rgb_colors = cmap(norm(norm_angles))

    return rgb_colors


def run_gui(history):
    num_steps = len(history)

    fig, ax = plt.subplots(figsize=(8, 8))
    plt.subplots_adjust(bottom=0.15)

    scatter = ax.scatter([], [], facecolors=[], cmap="viridis", vmin=0, vmax=1, s=10, edgecolor='k', linewidth=0)
    ax.set_xlim(-radius, radius)
    ax.set_ylim(-radius, radius)
    ax.set_aspect("equal")
    ax.set_title("Cell Simulation")
    
    # Draw boundary circle
    theta = np.linspace(0, 2*np.pi, 100)
    ax.plot(radius*np.cos(theta), radius*np.sin(theta), 'k--', linewidth=1)

    # Colorbar
    norm = mcolors.Normalize(vmin=0, vmax=1)
    sm = plt.cm.ScalarMappable(cmap="viridis", norm=norm)
    sm.set_array([])
    cbar = fig.colorbar(sm, ax=ax)
    cbar.set_label("phase")

    # Slider
    ax_slider = plt.axes([0.15, 0.05, 0.6, 0.03])
    slider = Slider(ax_slider, 'Timestep', 0, num_steps - 1, valinit=0, valstep=1)

    # Play/Pause Button
    playing = [False]  # wrapped in list to mutate from inner function
    ax_button = plt.axes([0.78, 0.05, 0.1, 0.04])
    button = Button(ax_button, "Play")

    def update(frame):
        data = history[frame]
        x = data[:, 0]
        y = data[:, 1]
        dx = data[:, 2]
        dy = data[:, 3]
        p = data[:, 4]
        # d = data[:, 6]

        # Compute angle in radians [-π, π], then normalize to [0, 1) for hue
        rgb_colors = skew_colors(x, y, dx, dy) 

        # Update scatter plot
        scatter.set_offsets(np.c_[x, y])
        # scatter.set_facecolors(rgb_colors)
        scatter.set_array(p)
        scatter.set_edgecolor('k')
        fig.canvas.draw_idle()


    def on_slider_change(val):
        update(int(val))

    def on_button_clicked(event):
        playing[0] = not playing[0]
        button.label.set_text("Pause" if playing[0] else "Play")

    def playback_loop(event):
        if playing[0]:
            current = int(slider.val)
            if current < num_steps - 1:
                slider.set_val(current + 1)
            else:
                slider.set_val(0)

    slider.on_changed(on_slider_change)
    button.on_clicked(on_button_clicked)

    # Timer for auto-play
    timer = fig.canvas.new_timer(interval=1)
    timer.add_callback(playback_loop, None)
    timer.start()

    update(0)
    plt.show()


# MARK: Trajectories

def plot_trajectories(history):
    # Determine the max number of cells
    max_cells = max(frame.shape[0] for frame in history)
    num_frames = len(history)

    # Initialize trajectory storage
    x_traj = [[] for _ in range(max_cells)]
    y_traj = [[] for _ in range(max_cells)]

    # Fill in trajectories
    for frame in history:
        x = frame[:, 0]
        y = frame[:, 1]
        for i in range(len(x)):
            x_traj[i].append(x[i])
            y_traj[i].append(y[i])

    # Plot trajectories
    fig, ax = plt.subplots(figsize=(20, 20))
    for xt, yt in zip(x_traj, y_traj):
        if len(xt) > 1:
            ax.plot(xt, yt, linewidth=0.8, alpha=0.6)

    # Draw boundary
    theta = np.linspace(0, 2*np.pi, 200)
    ax.plot(radius * np.cos(theta), radius * np.sin(theta), 'k--', linewidth=1)

    ax.set_title("Cell Trajectories")
    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.set_xlim(-radius, radius)
    ax.set_ylim(-radius, radius)
    ax.set_aspect('equal')
    plt.tight_layout()
    plt.savefig('trajectories.png')

plot_trajectories(history)
# Launch GUI
run_gui(history)

all_data = []
for t, frame in enumerate(history):
    df = pd.DataFrame(frame, columns=["x", "y", "dx", "dy", "p", "dp", "d"])
    df["timestep"] = t  # Add timestep info
    all_data.append(df)

full_df = pd.concat(all_data, ignore_index=True)
full_df.to_csv("cell_simulation_history.csv", index=False)